src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/VirtualDispatchHelpers.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.InteropServices;
+
+namespace System.Runtime.CompilerServices;
+
+[StackTraceHidden]
+[DebuggerStepThrough]
+internal static unsafe partial class VirtualDispatchHelpers
+{
+    private struct VirtualResolutionData : IEquatable<VirtualResolutionData>
+    {
+        public MethodTable* MethodTable;
+        public IntPtr ClassHandleTargetMethod;
+        public IntPtr MethodHandle;
+
+        public bool Equals(VirtualResolutionData other) =>
+            MethodTable == other.MethodTable &&
+            ClassHandleTargetMethod == other.ClassHandleTargetMethod &&
+            MethodHandle == other.MethodHandle;
+
+        public override bool Equals(object? obj) => obj is VirtualResolutionData other && Equals(other);
+
+        public override int GetHashCode() => (int) ((uint)MethodTable ^ (BitOperations.RotateLeft((uint)ClassHandleTargetMethod, 5)) ^ (BitOperations.RotateRight((uint)MethodHandle, 5)));
+    }
+
+    private struct VirtualFunctionPointerArgs
+    {
+        public IntPtr classHnd;
+        public IntPtr methodHnd;
+    };
+
+    private static Dictionary<VirtualResolutionData, IntPtr> s_virtualFunctionPointerCache = new Dictionary<VirtualResolutionData, IntPtr>();
+
+    internal static void ClearCache()
+    {
+        lock(s_virtualFunctionPointerCache)
+        {
+            s_virtualFunctionPointerCache.Clear();
+        }
+    }
+
+    [LibraryImport(RuntimeHelpers.QCall)]
+    private static unsafe partial IntPtr JIT_ResolveVirtualFunctionPointer(ObjectHandleOnStack obj, IntPtr classHandle, IntPtr methodHandle);
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    [DebuggerHidden]
+    private static unsafe IntPtr VirtualFunctionPointerSlowpath(object obj, IntPtr classHandle, IntPtr methodHandle)
+    {
+        IntPtr result = JIT_ResolveVirtualFunctionPointer(ObjectHandleOnStack.Create(ref obj), classHandle, methodHandle);
+        lock(s_virtualFunctionPointerCache)
```

Comment:
Regular locks can call back into user code through SynchronizationManager. It would be better for this to use Lock with `useTrivialWaits=true`.